openTSNE/tsne.py

```diff
@@ -1395,11 +1404,17 @@ def prepare_initial(self, X=None, affinities=None, initialization=None):
             raise ValueError(
                 f"Unrecognized initialization scheme `{initialization}`."
             )
+            
+        # Set the auto learning rate depending on the value of early exaggeration
+        if self.learning_rate == "auto":
+            learning_rate_now = max(200, n_samples / self.early_exaggeration)
+        else:
+            learning_rate_now = self.learning_rate 
 
         gradient_descent_params = {
             "dof": self.dof,
             "negative_gradient_method": self.negative_gradient_method,
-            "learning_rate": self.learning_rate,
+            "learning_rate": learning_rate_now,
```

Comment:
I'm not sure this is the correct put to put this. This would set the learning rate for the duration of the whole embedding, not just the early exaggeration phase, right? I think the best place to put this might actually be inside the `gradient_descent` call function, so the correct rescaling will happen during any call to `.optimize`, not just the standard `TSNE().fit` call.

I wasn't aware that this was an issue during the early exaggeration phase at all. It is actually necessary to rescale the learning rate, if the exaggeration isn't 1? In that case, wouldn't it also make sense to rescale it given _any_ exaggeration value, not just the early exaggeration phase?

Reply:
The current default openTSNE behaviour (which is the same in FIt-SNE and about to become default in soon-to-be-released sklearn 1.2) is to use learning rate N/12. Here 12 corresponds to the early exaggeration, the idea being that if learning_rate * exaggeration > 1, then the optimization may diverge (as shown by Linderman & Steinerberger). One COULD use learning rate N, instead of N/12, for the 2nd optimization phase, once early exaggeration is turned off. But we do not do it. We could decide to do it, and I think it may even be a good idea, but this will make openTSNE default behaviour different from other implementations. In any case, this I think requires a different PR and a separate issue, maybe comparing current and suggested scheme on a range of datasets.

In this PR I assumed that we want to keep the learning rate constant throughout optimization. Which is why I set it to N/early_exaggeration and keep the same learning rate for the second phase.

Reply:
Yes, that makes sense, I'd forgotten about that. However, I'm a bit unsure what the best way about implementing would be. We currently handle `learning_rate="auto"` [here](https://github.com/pavlin-policar/openTSNE/blob/master/openTSNE/tsne.py#L90-L92) in the `_handle_nice_params` function. However, this function is called in every `.optimize` step, so that we can perform whatever optimization sequence we want.

With the way you've implemented it now, we'd be going around this function because we'd change the `learning_rate` from `auto` to a number before we even got to this function. This would also cause inconsistent behaviour between
```python
TSNE().fit()
```
and
```python
embedding.optimize(...)
embedding.optimize(...)
```
since the first one would then use the "correct" learning rate, while the second would use `N/12` in both cases. I am very much opposed to this inconsistency, but I'm not sure what the best way to handle this would be.

Perhaps the best course of action would be to investigate rescaling the learning rate based on exaggeration, so we'd end up with a different learning rate for the early exaggeration and the standard optimization phase. I think this would best fit into our framework, and it seems the most principled approach.

Reply:
I see your point. I will start a separate issue.